File: thesis/algorithms/java/NpCompleto.java
import java.io.*;
import java.util.*;
import java.nio.file.Files;

class Item {
    int weight;
    int value;
    Item(int weight, int value) {
        this.weight = weight;
        this.value = value;
    }
}

public class NpCompleto {
    // Algoritmo guloso: ordena por valor/peso e adiciona enquanto possível
    public static int knapsack(List<Item> items, int capacity) {
        items.sort((a, b) -> Double.compare(
            (double)b.value / b.weight,
            (double)a.value / a.weight
        ));

        int totalValue = 0;
        int currentWeight = 0;

        for (Item item : items) {
            if (currentWeight + item.weight <= capacity) {
                currentWeight += item.weight;
                totalValue += item.value;
            }
        }

        return totalValue;
    }

    public static List<Item> readItems(String path, int[] capacityOut) throws IOException {
        String content = new String(Files.readAllBytes(new File(path).toPath()));
        int capacity = Integer.parseInt(content.split("\"capacity\":")[1].split(",")[0].trim());
        capacityOut[0] = capacity;
        List<Item> items = new ArrayList<>();
        String[] parts = content.split("\\{");
        for (String part : parts) {
            if (part.contains("weight") && part.contains("value")) {
                int weight = Integer.parseInt(part.split("weight\":")[1].split(",")[0].trim());
                int value = Integer.parseInt(part.split("value\":")[1].split("}")[0].trim());
                items.add(new Item(weight, value));
            }
        }
        return items;
    }

    public static void main(String[] args) throws IOException {
        String size = args.length > 0 ? args[0] : "small";
        String path = "datasets/" + size + "/knapsack.json";
        int[] capacity = new int[1];
        List<Item> items = readItems(path, capacity);
        int result = knapsack(items, capacity[0]);
        System.out.println("Valor aproximado (greedy) para " + items.size() +
                           " itens (capacidade " + capacity[0] + ", " + size + "): " + result);
    }
}

File: thesis/algorithms/c#/np-completo.cs
using System;
using System.Collections.Generic;
using System.IO;

class KnapsackItem {
    public int weight;
    public int value;
}

class KnapsackProgram {
    // Algoritmo guloso: ordena por valor/peso e adiciona enquanto possível
    static int Knapsack(List<KnapsackItem> items, int capacity) {
        items.Sort((a, b) => {
            double ra = (double)b.value / b.weight;
            double rb = (double)a.value / a.weight;
            return ra.CompareTo(rb);
        });

        int totalValue = 0;
        int currentWeight = 0;

        foreach (var item in items) {
            if (currentWeight + item.weight <= capacity) {
                currentWeight += item.weight;
                totalValue += item.value;
            }
        }

        return totalValue;
    }

    static void Main(string[] args) {
        string size = args.Length > 0 ? args[0] : "small";
        string path = $"datasets/{size}/knapsack.json";

        if (!File.Exists(path)) {
            Console.WriteLine("Arquivo não encontrado.");
            return;
        }

        string json = File.ReadAllText(path);

        // Obter capacidade
        int capacity = 0;
        int capIdx = json.IndexOf("\"capacity\"");
        if (capIdx != -1) {
            int start = json.IndexOf(":", capIdx) + 1;
            int end = json.IndexOf(",", start);
            string capStr = json.Substring(start, end - start).Trim();
            int.TryParse(capStr, out capacity);
        }

        // Obter itens
        List<KnapsackItem> items = new List<KnapsackItem>();
        int idx = 0;
        while ((idx = json.IndexOf("{\"weight\"", idx)) != -1) {
            int wStart = json.IndexOf(":", idx) + 1;
            int wEnd = json.IndexOf(",", wStart);
            string wStr = json.Substring(wStart, wEnd - wStart).Trim();

            int vStart = json.IndexOf(":", wEnd) + 1;
            int vEnd = json.IndexOf("}", vStart);
            string vStr = json.Substring(vStart, vEnd - vStart).Trim();

            if (int.TryParse(wStr, out int weight) && int.TryParse(vStr, out int value)) {
                items.Add(new KnapsackItem { weight = weight, value = value });
            }

            idx = vEnd + 1;
        }

        if (items.Count == 0) {
            Console.WriteLine("Erro ao ler o arquivo ou nenhum item encontrado.");
            return;
        }

        int result = Knapsack(items, capacity);
        Console.WriteLine($"Valor aproximado (greedy) para {items.Count} itens (capacidade {capacity}, {size}): {result}");
    }
}
